### PR TITLE
EDU-816 Create MusicXML plugin

### PR DESCRIPTION
**Pull request by ahelmberger**
Closes https://educandu.atlassian.net/browse/EDU-816

mxml

### CODE REVIEW

File: src/plugins/music-xml-viewer/music-xml-viewer.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+name:
+  en: MusicXML viewer
+  de: MusicXML-Anzeige
+zoom:
+  en: Zoom
+  de: Zoom
+caption:
+  en: Caption
+  de: Legende
```
Comment (adrianaluchian) [inline src/plugins/music-xml-viewer/music-xml-viewer.yml:9]: Tiny thing: the `pdf-viewer` plugin also has an identical "caption" translation, we could move it to `common.yml`

File: src/plugins/music-xml-viewer/music-xml-viewer-display.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    }
+  }, [actualUrl, zoom, osmd]);
+
+  return (
+    <div className="MusicXmlViewer">
+      <div className={`MusicXmlViewer-viewer u-width-${width || 100}`}>
+        <div ref={divRef} />
+      </div>
+      {!!caption && (
+        <div className={`PdfViewer-caption u-width-${width || 100}`}>
```
Comment (adrianaluchian) [inline src/plugins/music-xml-viewer/music-xml-viewer-display.js:52]: `MusicXmlViewer-caption`

File: src/plugins/music-xml-viewer/music-xml-viewer.less [hunk 3]
```diff
@@ -0,0 +1,12 @@
+.MusicXmlViewer {}
+
+.MusicXmlViewer-viewer {
+  margin: 0 auto;
+}
+
+.MusicXmlViewer-caption {
```
Comment (adrianaluchian) [inline src/plugins/music-xml-viewer/music-xml-viewer.less:7]: All these 3 classes could safely be called  `MusicXmlViewerDisplay[-viewer/caption]` as they are only used in the display component.

**Comment by adrianaluchian** [comment]
When I try to add a `music-xml-viewer` section in a document I get the following error:
Screen Shot 2022-09-28 at 09 34 39
Is it possible that you have some temporary local changes in this external package so that it works for you?

**Comment by ahelmberger** [comment]
> When I try to add a `music-xml-viewer` section in a document I get the following error: Screen Shot 2022-09-28 at 09 34 39 Is it possible that you have some temporary local changes in this external package so that it works for you?

`clear` seems to throw when there was no rendering before (and I already had test sections, so it never came to my attention). Unfortunately the whole library is not very nicely documented. I will re-write his a little bit.

File: src/plugins/music-xml-viewer/music-xml-viewer.yml (hunk 1 above)
Comment (ahelmberger) [reply src/plugins/music-xml-viewer/music-xml-viewer.yml]: Done!

File: src/plugins/music-xml-viewer/music-xml-viewer-info.spec.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import MusicXmlViewerInfo from './music-xml-viewer-info.js';
+
+describe('music-xml-viewer-info', () => {
+  let sut;
+  beforeEach(() => {
+    sut = new MusicXmlViewerInfo();
+  });
+
+  describe('redactContent', () => {
+    it('redacts the PDF source url', () => {
```
[the remark this answers is not in the document]
Comment (ahelmberger) [reply src/plugins/music-xml-viewer/music-xml-viewer-info.spec.js]: Darn copy/paste, it's "XML" now :laughing:

File: src/plugins/music-xml-viewer/music-xml-viewer-display.js (hunk 2 above)
Comment (ahelmberger) [reply src/plugins/music-xml-viewer/music-xml-viewer-display.js]: Thanks for spotting this one too!

File: src/plugins/music-xml-viewer/music-xml-viewer.less (hunk 3 above)
Comment (ahelmberger) [reply src/plugins/music-xml-viewer/music-xml-viewer.less]: Done (and also done for the PdfViewer[Display], which is almost the same plugin)!

**Comment by ahelmberger** [comment]
Sorry for the many copy/paste errors, I was very sloppy :roll_eyes:

File: src/plugins/music-xml-viewer/music-xml-viewer.less (hunk 3 above)
Comment (adrianaluchian) [reply src/plugins/music-xml-viewer/music-xml-viewer.less]: Poifect 👍

**Comment by adrianaluchian** [comment]
> Sorry for the many copy/paste errors, I was very sloppy 🙄

It's the project season for such things from both of us, all good!

**Comment by ahelmberger** [comment]
> > When I try to add a `music-xml-viewer` section in a document I get the following error: Screen Shot 2022-09-28 at 09 34 39 Is it possible that you have some temporary local changes in this external package so that it works for you?
> 
> `clear` seems to throw when there was no rendering before (and I already had test sections, so it never came to my attention). Unfortunately the whole library is not very nicely documented. I will re-write this a little bit.

I've extracted the component that renders the XML document (similar to how we've done it with PDF documents) and added some conditional logic to only clear when there was at least one render call before. Now the error is gone. Thanks for pointing it out.

File: src/components/pages/tests.js [hunk 5]
```diff
@@ -20,6 +20,7 @@ import MultitrackMediaEditor from '../../plugins/multitrack-media/multitrack-med
 import MultitrackMediaDisplay from '../../plugins/multitrack-media/multitrack-media-display.js';
 import { HORIZONTAL_ALIGNMENT, MEDIA_SCREEN_MODE, MEDIA_SOURCE_TYPE, STORAGE_LOCATION_TYPE, VERTICAL_ALIGNMENT } from '../../domain/constants.js';
 import { createDefaultContent, createDefaultMainTrack, createDefaultSecondaryTrack } from '../../plugins/multitrack-media/multitrack-media-utils.js';
+import MusicXmlDocument from '../music-xml-document.js';
```
Comment (adrianaluchian) [inline src/components/pages/tests.js:23]: These imports are staging a coup here 😆

**Comment by adrianaluchian** [comment]
> > > When I try to add a `music-xml-viewer` section in a document I get the following error: Screen Shot 2022-09-28 at 09 34 39 Is it possible that you have some temporary local changes in this external package so that it works for you?
> > 
> > 
> > `clear` seems to throw when there was no rendering before (and I already had test sections, so it never came to my attention). Unfortunately the whole library is not very nicely documented. I will re-write this a little bit.
> 
> I've extracted the component that renders the XML document (similar to how we've done it with PDF documents) and added some conditional logic to only clear when there was at least one render call before. Now the error is gone. Thanks for pointing it out.

Wonderful, looks great!! 🎉

**Review by adrianaluchian — APPROVED** [review]

File: src/components/pages/tests.js (hunk 5 above)
Comment (ahelmberger) [reply src/components/pages/tests.js]: :scream: I brought them to reason!

**Pull request merged**